fix(rootfs): create new recovery baseline after restore

bacongobbler opened the pull request:
When you complete a recovery of the database for the first time, a new log timeline
is started with an ID of 2. When you restore again (timeline of 3), the timeline used when the last
backup occurred will be replayed. Because of this, if you restored the database and
did not perform a backup, all data committed after that successful recovery will be lost because
only WAL logs from the first timeline (the timeline that the database was last backed
up) will be restored.

In order to fix this, after completing a database recovery we create a fresh backup
in order to establish a new recovery baseline. That way we can now replay from
timeline 2.

STEPS TO TEST:
1. deploy a workflow cluster with this version of postgres
2. `deis create` a few apps
3. "delete" (via `kubectl delete po`) the database to force a restart
4. repeat steps 2 and 3
5. observe that all applications are present, and not just the first few created.

closes #94

bacongobbler commented:
manually tested; works for me!

krancour commented:
@bacongobbler during manual testing, these are logs from the second attempt to restore from backup (first was successful):

```
$ k logs -f deis-database-aacbk
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
creating template1 database in /var/lib/postgresql/data/base/1 ... ok
initializing pg_authid ... ok
initializing dependencies ... ok
creating system views ... ok
loading system objects' descriptions ... ok
creating collations ... ok
creating conversions ... ok
creating dictionaries ... ok
setting privileges on built-in objects ... ok
creating information schema ... ok
loading PL/pgSQL server-side language ... ok
vacuuming database template1 ... ok
copying template1 to template0 ... ok
copying template1 to postgres ... ok
syncing data to disk ... ok

WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    postgres -D /var/lib/postgresql/data
or
    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....LOG:  database system was shut down at 2016-05-27 19:00:50 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
CREATE DATABASE

CREATE ROLE


/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/001_setup_envdir.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/002_create_bucket.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/003_restore_from_backup.sh
Rebooting postgres to enable archive mode
waiting for server to shut down...LOG:  received smart shutdown request
.LOG:  autovacuum launcher shutting down
LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
waiting for server to start....LOG:  database system was shut down at 2016-05-27 19:00:52 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
Found backups. Restoring from backup...
waiting for server to shut down....LOG:  received smart shutdown request
LOG:  autovacuum launcher shutting down
LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-fetch".
        STRUCTURED: time=2016-05-27T19:00:56.068343-00 pid=105
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.135363-00 pid=105
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.3.0.227
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.197708-00 pid=105
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.202305-00 pid=105
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_00000002000000000000000A_00000040_backup_stop_sentinel.json'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.3.0.227
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.415208-00 pid=105
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.427510-00 pid=105
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.433979-00 pid=105
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:00:56.440143-00 pid=105
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/base_00000002000000000000000A_00000040/tar_partitions/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.3.0.227
wal_e.worker.s3.s3_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute S3 key is basebackups_005/base_00000002000000000000000A_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2016-05-27T19:00:56.571725-00 pid=105
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_00000002000000000000000A_00000040/tar_partitions/part_00000000.tar.lzo'}
waiting for server to start....LOG:  database system was interrupted; last known up at 2016-05-27 18:58:19 UTC
LOG:  creating missing WAL directory "pg_xlog/archive_status"
LOG:  starting archive recovery
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-05-27T19:01:15.693152-00 pid=125
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-05-27T19:01:15.767577-00 pid=125 action=wal-fetch key=s3://dbwal/wal_005/00000002.history.lzo prefix= seg=00000002.history state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:01:15.868540-00 pid=125
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/00000002.history.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.3.0.227
lzop: <stdin>: not a lzop file
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/00000002.history.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-05-27T19:01:16.074832-00 pid=125
FATAL:  the database system is starting up
.wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-05-27T19:01:16.176260-00 pid=125 action=wal-fetch key=s3://dbwal/wal_005/00000002.history.lzo prefix= seg=00000002.history state=complete
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-05-27T19:01:16.406688-00 pid=136
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-05-27T19:01:16.487151-00 pid=136 action=wal-fetch key=s3://dbwal/wal_005/00000002000000000000000A.lzo prefix= seg=00000002000000000000000A state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.3.0.227:9000)
        STRUCTURED: time=2016-05-27T19:01:16.589366-00 pid=136
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/00000002000000000000000A.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.3.0.227
lzop: <stdin>: not a lzop file
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/00000002000000000000000A.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-05-27T19:01:17.995903-00 pid=136
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-05-27T19:01:18.097744-00 pid=136 action=wal-fetch key=s3://dbwal/wal_005/00000002000000000000000A.lzo prefix= seg=00000002000000000000000A state=complete
LOG:  invalid checkpoint record
FATAL:  could not locate required checkpoint record
HINT:  If you are not restoring from a backup, try removing the file "/var/lib/postgresql/data/backup_label".
FATAL:  the database system is starting up
.LOG:  startup process (PID 123) exited with exit code 1
LOG:  aborting startup due to startup process failure
pg_ctl: could not start server
Examine the log output.
 stopped waiting
```

krancour commented:
I wiped everything out and started fresh and still see that same error on every second restore attempt.

bacongobbler commented:
that seems to be related, yet also unrelated because others have hit this too: https://github.com/wal-e/wal-e/issues/251 https://github.com/wal-e/wal-e/issues/189

I'll see if I can dig deeper into this problem to see if there's any fixes required upstream.

arschles commented:
@bacongobbler are there any specific manual testing steps needed here?

bacongobbler commented:
specifically:
1. deploy a workflow cluster with this version of postgres
2. `deis create` a few apps
3. "delete" (via `kubectl delete po`) the database to force a restart
4. repeat steps 2 and 3
5. observe that all applications are present, and not just the first few created.

bacongobbler commented:
I'll see if I can write some integration tests for this as well

arschles commented:
@bacongobbler I completed steps 1-3 (on GKE, with Minio object storage - otherwise known as the default workflow-dev installation) with the following commands:

``` console
ENG000656:postgres aaronschlesinger$ deis create --no-remote app1 && deis create --no-remote app2 && deis create --no-remote app3
Creating Application... done, created app1
remote available at ssh://git@deis.104.197.168.155.nip.io:2222/app1.git
Creating Application... done, created app2
remote available at ssh://git@deis.104.197.168.155.nip.io:2222/app2.git
Creating Application... done, created app3
remote available at ssh://git@deis.104.197.168.155.nip.io:2222/app3.git
ENG000656:postgres aaronschlesinger$ kd delete po deis-database-mphfo
pod "deis-database-mphfo" deleted
```

... and the new DB pod wouldn't start up. Here are the logs for it:

``` console
^CENG000656:postgres aaronschlesinger$ kd logs -f deis-database-rs1w0
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
creating template1 database in /var/lib/postgresql/data/base/1 ... ok
initializing pg_authid ... ok
initializing dependencies ... ok
creating system views ... ok
loading system objects' descriptions ... ok
creating collations ... ok
creating conversions ... ok
creating dictionaries ... ok
setting privileges on built-in objects ... ok
creating information schema ... ok
loading PL/pgSQL server-side language ... ok
vacuuming database template1 ... ok
copying template1 to template0 ... ok
copying template1 to postgres ... ok
syncing data to disk ... ok

Success. You can now start the database server using:

    postgres -D /var/lib/postgresql/data
or
    pg_ctl -D /var/lib/postgresql/data -l logfile start


WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.
waiting for server to start....LOG:  database system was shut down at 2016-06-01 18:49:53 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
CREATE DATABASE

CREATE ROLE


/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/001_setup_envdir.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/002_create_bucket.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/003_restore_from_backup.sh
Rebooting postgres to enable archive mode
waiting for server to shut down...LOG:  received smart shutdown request
.LOG:  autovacuum launcher shutting down
LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
waiting for server to start....LOG:  database system was shut down at 2016-06-01 18:50:09 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
Found backups. Restoring from backup...
waiting for server to shut down...LOG:  received smart shutdown request
LOG:  autovacuum launcher shutting down
.LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-fetch".
        STRUCTURED: time=2016-06-01T18:50:13.551001-00 pid=104
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.648922-00 pid=104
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.240.185
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.748616-00 pid=104
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.753725-00 pid=104
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_000000010000000000000002_00000040_backup_stop_sentinel.json'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.240.185
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.771295-00 pid=104
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.778227-00 pid=104
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.782677-00 pid=104
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:13.787203-00 pid=104
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/base_000000010000000000000002_00000040/tar_partitions/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.240.185
wal_e.worker.s3.s3_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute S3 key is basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2016-06-01T18:50:13.908649-00 pid=104
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo'}
waiting for server to start....LOG:  database system was interrupted; last known up at 2016-06-01 18:44:45 UTC
LOG:  creating missing WAL directory "pg_xlog/archive_status"
FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.LOG:  starting archive recovery
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-01T18:50:29.481567-00 pid=149
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-01T18:50:29.595448-00 pid=149 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000002.lzo prefix= seg=000000010000000000000002 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.240.185:9000)
        STRUCTURED: time=2016-06-01T18:50:29.720265-00 pid=149
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000002.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.240.185
lzop: <stdin>: not a lzop file
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/000000010000000000000002.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-06-01T18:50:30.682669-00 pid=149
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-01T18:50:30.784082-00 pid=149 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000002.lzo prefix= seg=000000010000000000000002 state=complete
FATAL:  the database system is starting up
.LOG:  invalid checkpoint record
FATAL:  could not locate required checkpoint record
HINT:  If you are not restoring from a backup, try removing the file "/var/lib/postgresql/data/backup_label".
LOG:  startup process (PID 124) exited with exit code 1
LOG:  aborting startup due to startup process failure
 stopped waiting
pg_ctl: could not start server
Examine the log output.
```

Note that I built the image myself on your branch with `make all`. Hopefully I didn't miss anything there...

bacongobbler commented:
I did not see this on my second run-though, however since both of you have hit this I'm going to continue to research into this to ensure I'm not breaking anything :)

logs:

```
><> deis create go --no-remote
Creating Application... done, created go
If you want to add a git remote for this app later, use `deis git:remote -a go`
><> deis apps
=== Apps
go
><> kd delete po deis-database-e3l7b
pod "deis-database-e3l7b" deleted
><> sleep 120
><> deis apps
=== Apps
go
><> deis create foo --no-remote
Creating Application... done, created foo
If you want to add a git remote for this app later, use `deis git:remote -a foo`
><> kd delete po deis-database-due69          
pod "deis-database-due69" deleted
><> sleep 120
><> deis apps
=== Apps
go
foo
```

EDIT: perhaps https://github.com/deis/postgres/issues/55 could be a culprit, but like I said, more research is needed.
EDIT2: gone through 3 recoveries without failure now, so it seems unrelated.

bacongobbler commented:
I'm going to turn this into a (slightly) larger refactor that will hopefully alleviate the symptoms that caused this checkpoint failure. From reading up on the issue as well as looking at the logs, it looks like the database is attempting to start while it is still halfway through recovery mode, which @krancour and I tried to alleviate earlier with #59. While it is not directly related to the original change made in this PR (and I can separate these two into separate PRs if you wish), but it will greatly improve the stability of database recovery because we will only be booting the database once when restoring from a backup rather than twice. I'm just manually testing this, but if anyone feels adventurous today you're more than welcome to test as well. :)

bacongobbler commented:
Moving this to "in-progress" until I can see that this refactor is ready for review again.

slack commented:
Hmm. What does `gosu postgres pg_ctl status` do when the database is recovering?

bacongobbler commented:
it'll say that the database is not ready to accept incoming connections. How come?

arschles commented:
@slack was that a question for me?

slack commented:
Sorry, just thinking out loud. In https://github.com/deis/postgres/issues/55 the database is ostensibly still recovering from backups but is showing as `Ready 1/1`. It seems that the readiness check `is_running` should be failing while in recovery, leaving the Pod as `0/1`.

Makes me wonder if our guard isn't really acting as guard, thus the question about `gosu postgres pg_ctl status` during recovery.

bacongobbler commented:
Yeah actually I found that my statement is false. I'm fixing that right now as well as I saw it during my testing, so hopfully we should be able to close #55 with this as well:

```
PostgreSQL init process complete; ready for start up.

LOG:  database system was interrupted; last known up at 2016-06-01 21:58:35 UTC
LOG:  creating missing WAL directory "pg_xlog/archive_status"
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-push".
        STRUCTURED: time=2016-06-01T22:03:52.721883-00 pid=115
FATAL:  the database system is starting up
psql: FATAL:  the database system is starting up
wal_e.main   ERROR    MSG: could not run one or more external programs WAL-E depends upon
        DETAIL: Could not get a connection to the database: note that superuser access
        is required
        Could not run the following programs, are they installed? psql
        STRUCTURED: time=2016-06-01T22:03:52.846007-00 pid=115
```

bacongobbler commented:
works now!

slack commented on rootfs/docker-entrypoint.sh:
```diff
@@ -80,29 +80,37 @@ if [ "$1" = 'postgres' ]; then
 		EOSQL
 		echo
 
+		gosu postgres pg_ctl -D "$PGDATA" -m fast -w stop
```
Assuming none of the `/docker-entrypoint-initdb.d` scripts require postgres to be running?

bacongobbler replied:
not any more! The commands that _do_ require postgres to be running are now run after the database has been booted and is running.

slack commented on rootfs/docker-entrypoint.sh:
```diff
@@ -80,29 +80,47 @@ if [ "$1" = 'postgres' ]; then
+	gosu postgres envdir "$WALE_ENVDIR" wal-e backup-push "$PGDATA"
+
+	# Run periodic backups in the background
+	gosu postgres backup &
```
Do you think we will need to trap SIGINT and SIGTERM in the backup process as well?

bacongobbler replied:
I don't think /bin/backup can handle SIGTERM gently at this time (due to /bin/backup being an infinite `while` loop), though that is certainly something we should definitely consider in the future. Keep in mind that v1 did not cleanly shut down /bin/backup either and we didn't see any issues in production since the time to backup is very short and only occurs once every 12 hours by default, so the potential window of loss is very minimal. However, I'd be happy to add a ticket that which we can test, observe and gracefully handle SIGTERM in the middle of a base backup in the future. :)

For reference, the SIGTERM handling code seen here [is code from Deis v1](https://github.com/deis/deis/blob/955dcea900d6822590c8bb01bb30d9fbf9190ac6/database/bin/boot#L103-L116) since postgres is now no longer PID 1 with this refactor. I'd love to get back to `exec`ing the main process so postgres can run as PID 1 for process management purposes, but given that we have multiple back-grounded processes running in the container that depend on each other (`postgres` and `/bin/backup`), I can't see a way around that. Without this crutch, the database would shut down immediately without forcing an immediate WAL log shipped to object storage, leaving us with a potential window of loss.

slack replied:
No problemo. Sounds like we don't need to do anything right now. We can deal with it later. May also want to think about moving this into a sidecar container in the pod.

bacongobbler replied:
brilliant idea! The only problem I see with that is the backup needs to run immediately after the database is started. That way we can ensure that we have a reliable base backup and that we don't have to replay everything from scratch again. Sounds like something we can address in that ticket though.

bacongobbler commented:
Finally back and ready for review once again. I'm currently manually testing this, but integration tests are passing once again.

bacongobbler commented:
haa, hit #67 again

bacongobbler commented:
right now I'm finding database corruption errors. not sure why, but investigating...

Commit:
fix(rootfs): always perform an initial backup

When you complete a recovery of the database for the first time, a new log timeline is started with
an ID of 2. When you restore again (timeline of 3), the timeline used when the last backup occurred
will be replayed. Because of this, if you restored the database and did not perform a backup, all
data committed after that successful recovery will be lost because only WAL logs from the first
timeline (the timeline that the database was last backed up) will be restored.

In order to fix this, after completing a database recovery we create a fresh backup
in order to establish a new recovery baseline. That way we can now replay from
timeline 2.

bacongobbler commented:
Leaving the refactors out of this PR because I cannot seem to fix the database corruption errors and is now blocking the RC2 release. In any case, this should fix the original issue. The checkpoint errors can be addressed in the future

bacongobbler commented:
refactors can be found at https://github.com/bacongobbler/postgres/tree/refactor-database

bacongobbler commented:
ci is currently blocked by deis/minio#107

arschles commented:
@bacongobbler I tested again on the `workflow-rc2` chart (using Minio as object storage) and found that the database comes back up after step 3, but then `deis apps:list` returns no apps. I also tried deleting the DB pod a second time, and got the same behavior.

bacongobbler commented:
@arschles would you mind sharing your database logs in a gist so we can go over the the recovery process?

arschles commented:
@bacongobbler sure! Below are the DB logs after the second pod deletion. Did you need the ones after the first pod deletion?

``` console
ENG000656:postgres aaronschlesinger$ kd logs -f deis-database-svz6p
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
creating template1 database in /var/lib/postgresql/data/base/1 ... ok
initializing pg_authid ... ok
initializing dependencies ... ok
creating system views ... ok
loading system objects' descriptions ... ok
creating collations ... ok
creating conversions ... ok
creating dictionaries ... ok
setting privileges on built-in objects ... ok
creating information schema ... ok
loading PL/pgSQL server-side language ... ok
vacuuming database template1 ... ok
copying template1 to template0 ... ok
copying template1 to postgres ... ok

WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.
syncing data to disk ... ok

Success. You can now start the database server using:

    postgres -D /var/lib/postgresql/data
or
    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....LOG:  database system was shut down at 2016-06-06 18:06:15 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
CREATE DATABASE

CREATE ROLE


/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/001_setup_envdir.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/002_create_bucket.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/003_restore_from_backup.sh
Rebooting postgres to enable archive mode
waiting for server to shut down...LOG:  received smart shutdown request
LOG:  autovacuum launcher shutting down
.LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
waiting for server to start....LOG:  database system was shut down at 2016-06-06 18:06:32 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
Found backups. Restoring from backup...
waiting for server to shut down...LOG:  received smart shutdown request
LOG:  autovacuum launcher shutting down
.LOG:  shutting down
LOG:  database system is shut down
 done
server stopped
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-fetch".
        STRUCTURED: time=2016-06-06T18:06:35.977512-00 pid=102
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.060837-00 pid=102
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.153253-00 pid=102
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.158504-00 pid=102
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_000000010000000000000002_00000040_backup_stop_sentinel.json'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.178073-00 pid=102
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.183580-00 pid=102
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.187753-00 pid=102
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:36.191724-00 pid=102
boto3.resources.collection INFO     Calling paginated s3:list_objects with {'Prefix': 'basebackups_005/base_000000010000000000000002_00000040/tar_partitions/', u'Bucket': 'dbwal'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.worker.s3.s3_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute S3 key is basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2016-06-06T18:06:36.313862-00 pid=102
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo'}
waiting for server to start....LOG:  database system was interrupted; last known up at 2016-06-06 17:56:28 UTC
LOG:  creating missing WAL directory "pg_xlog/archive_status"
FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.LOG:  starting archive recovery
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:06:51.790737-00 pid=136
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:06:51.883755-00 pid=136 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000002.lzo prefix= seg=000000010000000000000002 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:51.985904-00 pid=136
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000002.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://dbwal/wal_005/000000010000000000000002.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2016-06-06T18:06:53.072570-00 pid=136
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:06:53.073484-00 pid=136 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000002.lzo prefix= seg=000000010000000000000002 state=complete
LOG:  restored log file "000000010000000000000002" from archive
FATAL:  the database system is starting up
.LOG:  redo starts at 0/2000090
LOG:  consistent recovery state reached at 0/20000B8
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:06:54.261426-00 pid=180
FATAL:  the database system is starting up
.wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:06:54.552852-00 pid=180 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000003.lzo prefix= seg=000000010000000000000003 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:54.660278-00 pid=180
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000003.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://dbwal/wal_005/000000010000000000000003.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2016-06-06T18:06:55.995691-00 pid=180
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:06:55.996523-00 pid=180 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000003.lzo prefix= seg=000000010000000000000003 state=complete
LOG:  restored log file "000000010000000000000003" from archive
FATAL:  the database system is starting up
.wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:06:57.030472-00 pid=196
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:06:57.185021-00 pid=196 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000004.lzo prefix= seg=000000010000000000000004 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:57.290338-00 pid=196
FATAL:  the database system is starting up
.botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000004.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://dbwal/wal_005/000000010000000000000004.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2016-06-06T18:06:58.072506-00 pid=196
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:06:58.073406-00 pid=196 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000004.lzo prefix= seg=000000010000000000000004 state=complete
LOG:  restored log file "000000010000000000000004" from archive
FATAL:  the database system is starting up
.wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:06:58.985842-00 pid=211
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:06:59.131684-00 pid=211 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000005.lzo prefix= seg=000000010000000000000005 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:06:59.234085-00 pid=211
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000005.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
FATAL:  the database system is starting up
.wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://dbwal/wal_005/000000010000000000000005.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2016-06-06T18:07:00.883393-00 pid=211
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:07:00.884206-00 pid=211 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000005.lzo prefix= seg=000000010000000000000005 state=complete
FATAL:  the database system is starting up
.LOG:  restored log file "000000010000000000000005" from archive
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:07:02.342059-00 pid=226
FATAL:  the database system is starting up
.wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:07:02.493353-00 pid=226 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000006.lzo prefix= seg=000000010000000000000006 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:07:02.595709-00 pid=226
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000006.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
lzop: <stdin>: not a lzop file
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/000000010000000000000006.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-06-06T18:07:04.950954-00 pid=226
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:07:05.063965-00 pid=226 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000006.lzo prefix= seg=000000010000000000000006 state=complete
FATAL:  the database system is starting up
.LOG:  redo done at 0/5000420
LOG:  last completed transaction was at log time 2016-06-06 17:59:30.666766+00
FATAL:  the database system is starting up
.wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:07:06.613647-00 pid=248
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:07:06.793223-00 pid=248 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000005.lzo prefix= seg=000000010000000000000005 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:07:06.896497-00 pid=248
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000010000000000000005.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
FATAL:  the database system is starting up
.wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://dbwal/wal_005/000000010000000000000005.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2016-06-06T18:07:07.592071-00 pid=248
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:07:07.592907-00 pid=248 action=wal-fetch key=s3://dbwal/wal_005/000000010000000000000005.lzo prefix= seg=000000010000000000000005 state=complete
LOG:  restored log file "000000010000000000000005" from archive
FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:07:09.534886-00 pid=265
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:07:09.844246-00 pid=265 action=wal-fetch key=s3://dbwal/wal_005/00000002.history.lzo prefix= seg=00000002.history state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:07:09.945722-00 pid=265
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/00000002.history.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/00000002.history.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-06-06T18:07:11.380397-00 pid=265
lzop: <stdin>: not a lzop file
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:07:11.482346-00 pid=265 action=wal-fetch key=s3://dbwal/wal_005/00000002.history.lzo prefix= seg=00000002.history state=complete
FATAL:  the database system is starting up
.LOG:  selected new timeline ID: 2
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2016-06-06T18:07:12.518530-00 pid=294
FATAL:  the database system is starting up
.wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2016-06-06T18:07:12.674991-00 pid=294 action=wal-fetch key=s3://dbwal/wal_005/00000001.history.lzo prefix= seg=00000001.history state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:07:12.776641-00 pid=294
botocore.credentials INFO     Found credentials in environment variables.
FATAL:  the database system is starting up
.boto3.resources.action INFO     Calling s3:get_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/00000001.history.lzo'}
FATAL:  the database system is starting up
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
.lzop: <stdin>: not a lzop file
wal_e.blobstore.s3.s3_util WARNING  MSG: could no longer locate object while performing wal restore
        DETAIL: The absolute URI that could not be located is s3://dbwal/wal_005/00000001.history.lzo.
        HINT: This can be normal when Postgres is trying to detect what timelines are available during restoration.
        STRUCTURED: time=2016-06-06T18:07:14.575524-00 pid=294
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2016-06-06T18:07:14.676685-00 pid=294 action=wal-fetch key=s3://dbwal/wal_005/00000001.history.lzo prefix= seg=00000001.history state=complete
LOG:  archive recovery complete
FATAL:  the database system is starting up
.FATAL:  the database system is starting up
.LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/004_setup_backup_restore.sh

/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/005_run_backups.sh

LOG:  received fast shutdown request
LOG:  aborting any active transactions
LOG:  autovacuum launcher shutting down
LOG:  shutting down
waiting for server to shut down....LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

LOG:  database system was shut down at 2016-06-06 18:07:17 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-push".
        STRUCTURED: time=2016-06-06T18:08:19.444137-00 pid=473
wal_e.worker.upload INFO     MSG: begin archiving a file
        DETAIL: Uploading "pg_xlog/000000020000000000000006" to "s3://dbwal/wal_005/000000020000000000000006.lzo".
        STRUCTURED: time=2016-06-06T18:08:19.542452-00 pid=473 action=push-wal key=s3://dbwal/wal_005/000000020000000000000006.lzo prefix= seg=000000020000000000000006 state=begin
wal_e.blobstore.s3.calling_format INFO     MSG: connecting to S3 with WALE_S3_ENDPOINT (http://10.171.244.247:9000)
        STRUCTURED: time=2016-06-06T18:08:19.706038-00 pid=473
botocore.credentials INFO     Found credentials in environment variables.
boto3.resources.action INFO     Calling s3:put_object with {'Body': <open file '<fdopen>', mode 'r+b' at 0x7ff58787dc00>, u'Bucket': 'dbwal', 'Key': 'wal_005/000000020000000000000006.lzo'}
botocore.vendored.requests.packages.urllib3.connectionpool INFO     Starting new HTTP connection (1): 10.171.244.247
boto3.resources.action INFO     Calling s3:head_object with {u'Bucket': 'dbwal', u'Key': 'wal_005/000000020000000000000006.lzo'}
wal_e.worker.upload INFO     MSG: completed archiving to a file
        DETAIL: Archiving to "s3://dbwal/wal_005/000000020000000000000006.lzo" complete at 798.766KiB/s.
        STRUCTURED: time=2016-06-06T18:08:19.813769-00 pid=473 action=push-wal key=s3://dbwal/wal_005/000000020000000000000006.lzo prefix= rate=798.766 seg=000000020000000000000006 state=complete
```

arschles commented:
@bacongobbler my previous comment was using the RC2 postgres image. Upon using an image created from this PR, things look good after both restarts. Logs below.

``` console
ENG000656:postgres aaronschlesinger$ deis create --no-remote app1
Creating Application... o..de.o.c..odone, created app1
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app1.git
ENG000656:postgres aaronschlesinger$ deis create --no-remote app2
Creating Application... done, created app2
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app2.git
ENG000656:postgres aaronschlesinger$ deis create --no-remote app3
Creating Application... done, created app3
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app3.git
ENG000656:postgres aaronschlesinger$ deis apps
=== Apps
app1
app2
app3
ENG000656:postgres aaronschlesinger$ kd delete po deis-database-d024c
pod "deis-database-d024c" deleted
ENG000656:postgres aaronschlesinger$ deis apps
=== Apps
app1
app2
app3
ENG000656:postgres aaronschlesinger$ deis create --no-remote app4
Creating Application... done, created app4
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app4.git
ENG000656:postgres aaronschlesinger$ deis create --no-remote app5
Creating Application... done, created app5
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app5.git
ENG000656:postgres aaronschlesinger$ deis create --no-remote app6
Creating Application... done, created app6
remote available at ssh://git@deis.104.197.220.194.nip.io:2222/app6.git
ENG000656:postgres aaronschlesinger$ kd delete po deis-database-ahnmh
pod "deis-database-ahnmh" deleted
ENG000656:postgres aaronschlesinger$ deis apps
=== Apps
app1
app2
app3
app4
app5
app6
```

bacongobbler commented:
Excellent! Thank you for testing. That recovery error is alarming since this does not fix the recovery process (it only creates a base backup _after_ recovery), but hopefully that will be resolved in my refactoring branch at https://github.com/bacongobbler/postgres/tree/refactor-database which I hope to land soon to get rid of these errors.

arschles commented:
@bacongobbler maybe I can add some detail to the recovery error in case it applies:
- I deleted the database pod almost immediately after I created the third app (`app3`)
- I was using the RC2 database at the time

bacongobbler commented:
that helps immensely. Thank you

bacongobbler commented:
note that #112 should help alleviate some of those issues you're seeing, since we only boot the database twice instead of three times.

slack commented:
Built and tested with image `jhansen/postgres:git-db4cd9d`.

Recovered twice with a base backup in between.

bacongobbler commented:
thanks @slack for the extra security blanket :)
